Minimum similarity bar in dendrogram

sehilyi opened the pull request:
- Added a condition called `minSimilarity` in filter/highlight options.
- Added a toggle button in dendrogram to show/hide a minimum similarity bar.
- Added ability to filter/highlight using the bar.
- Changed `onSearchRows` to `onHighlightRows` since highlighting can be now done not only with searching keywords.

![min-sim-bar](https://user-images.githubusercontent.com/9922882/82090824-485c8200-96c4-11ea-8beb-6a6016d04f18.gif)

> Gotta find for an alternative gif generator other than GIPHY CAPTURE. Hard to meet the file size limit when uploading.

(Fixes #228)

keller-mark commented:
It seems like sometimes I get an extra row after filtering: in first screenshot below, I moved the bar, and then second screenshot is after right-click and "Filter Rows"

![Screen Shot 2020-05-18 at 10 07 37 AM](https://user-images.githubusercontent.com/7525285/82222784-c781ce00-98ef-11ea-9d34-25787cad0811.png)

![Screen Shot 2020-05-18 at 10 07 42 AM](https://user-images.githubusercontent.com/7525285/82222796-ca7cbe80-98ef-11ea-973b-8d93663ccab6.png)

keller-mark commented:
It seems that filtering followed by highlighting may not use the subtree. I tried the following on the `Minimal Dataset (w/ Dendrogram Similarity Distance)` dataset

![Screen Shot 2020-05-18 at 10 47 53 AM](https://user-images.githubusercontent.com/7525285/82227042-41688600-98f5-11ea-94a8-d4ef9944efcd.png)

1. Filter by subtree: 

![Screen Shot 2020-05-18 at 10 48 00 AM](https://user-images.githubusercontent.com/7525285/82227166-63fa9f00-98f5-11ea-8126-bf3b0dd29957.png)

2. Move similarity bar
![Screen Shot 2020-05-18 at 10 48 08 AM](https://user-images.githubusercontent.com/7525285/82227193-6c52da00-98f5-11ea-9d2b-503d54c0d3f5.png)

3. Filter, expecting 4 rows, but got more than 4

![Screen Shot 2020-05-18 at 10 48 17 AM](https://user-images.githubusercontent.com/7525285/82227219-74ab1500-98f5-11ea-9cce-8d6fc5268a44.png)

keller-mark commented:
Can we add the `default.profraw` to the `.gitignore`?

keller-mark commented:
It looks like changing changing the width does not automatically update the subtree opacity. Can we detect when the width has changed and re-compute the opacities? 

![Screen Shot 2020-05-18 at 11 01 27 AM](https://user-images.githubusercontent.com/7525285/82228397-0ebf8d00-98f7-11ea-894d-28cb6ef71630.png)
![Screen Shot 2020-05-18 at 11 01 36 AM](https://user-images.githubusercontent.com/7525285/82228399-0f582380-98f7-11ea-973e-0ef295edc2c3.png)

sehilyi commented on src/utils/options.js:
```diff
@@ -31,7 +31,9 @@ const baseSchema = {
-                        { "required": ["field", "type", "range"] }
+                        { "required": ["field", "type", "range"] },
+                        { "required": ["field", "type", "subtree"] },
+                        { "required": ["field", "type", "minSimilarity"] },
```
Currently, this is not implemented because our interface do not support applying multiple highlighting options for individual fields in the interface (i.e., new highlight option replace the previous one). But, if we find this use case to be essential, I think we can implement that with a separate PR.

sehilyi commented on src/utils/select-rows.js:
```diff
@@ -32,10 +32,19 @@ export function selectRows(rowInfo, options) {
+                            (a, h, i) => a && (i >= subtree.length || h === subtree[i] || h.name === subtree[i]), true)
+                        );
+                    } 
+                    if (minSimilarity) {
```
> It seems that filtering followed by highlighting may not use the subtree. I tried the following on the `Minimal Dataset (w/ Dendrogram Similarity Distance)` dataset
> 
> ![Screen Shot 2020-05-18 at 10 47 53 AM](https://user-images.githubusercontent.com/7525285/82227042-41688600-98f5-11ea-94a8-d4ef9944efcd.png)
> 
> 1. Filter by subtree:
> 
> ![Screen Shot 2020-05-18 at 10 48 00 AM](https://user-images.githubusercontent.com/7525285/82227166-63fa9f00-98f5-11ea-8126-bf3b0dd29957.png)
> 
> 1. Move similarity bar
>    ![Screen Shot 2020-05-18 at 10 48 08 AM](https://user-images.githubusercontent.com/7525285/82227193-6c52da00-98f5-11ea-9d2b-503d54c0d3f5.png)
> 2. Filter, expecting 4 rows, but got more than 4
> 
> ![Screen Shot 2020-05-18 at 10 48 17 AM](https://user-images.githubusercontent.com/7525285/82227219-74ab1500-98f5-11ea-9cce-8d6fc5268a44.png)

I am trying to reproduce this case: In the "1. Move similarity bar", did you highlight the rows on the bottom using the context menu?

sehilyi commented:
> It seems that filtering followed by highlighting may not use the subtree. I tried the following on the `Minimal Dataset (w/ Dendrogram Similarity Distance)` dataset
> 
> ![Screen Shot 2020-05-18 at 10 47 53 AM](https://user-images.githubusercontent.com/7525285/82227042-41688600-98f5-11ea-94a8-d4ef9944efcd.png)
> 
> 1. Filter by subtree:
> 
> ![Screen Shot 2020-05-18 at 10 48 00 AM](https://user-images.githubusercontent.com/7525285/82227166-63fa9f00-98f5-11ea-8126-bf3b0dd29957.png)
> 
> 1. Move similarity bar
>    ![Screen Shot 2020-05-18 at 10 48 08 AM](https://user-images.githubusercontent.com/7525285/82227193-6c52da00-98f5-11ea-9d2b-503d54c0d3f5.png)
> 2. Filter, expecting 4 rows, but got more than 4
> 
> ![Screen Shot 2020-05-18 at 10 48 17 AM](https://user-images.githubusercontent.com/7525285/82227219-74ab1500-98f5-11ea-9cce-8d6fc5268a44.png)

I am trying to reproduce this case: In the "1. Move similarity bar", did you highlight the rows on the bottom using the context menu?

keller-mark commented:
Hmm I cannot seem to reproduce this either, using the instructions I just wrote. Maybe I changed something while reviewing the code, but I can filter by subtree and then similarity now! Sorry about that!

sehilyi commented:
Did the rows on the bottom highlighted by just moving the minimum similarity bar? I removed this functionality (i.e., highlighting rows by default with the bar movement) because of a known issue just like the one you've mentioned in the above comment.

keller-mark commented:
Yes I was trying to understand your TODO comment, that might have been it

sehilyi commented:
Oh, I got it. Thanks for the clarification. I will need to look into that line more.

sehilyi commented:
Thank you for finding these issue!

The first bug you've mentioned was occurred when a node has at least one branch that has a closer `dist` value than `minSimilarity` but has no nodes to be connected with (for example, by a certain filtering status). I've added an additional filtering process for this [here](https://github.com/hms-dbmi/cistrome-higlass-wrapper/blob/9a4094e0fc5fe2aeba1043183672b41ffaaca482/src/utils/select-rows.js#L51).

I've also resolved all other comments you've proposed except for the test. I think the test for `select-rows` can be added in a separate PR since I believe there is no problems in releasing the current `select-rows.js`. As it is definitely necessary, I will add a "high-priority" issue for this.

sehilyi commented:
Thank you for your constructive feedback and comments!

The first bug you've mentioned was occurred when a node has at least one branch that has a closer `dist` value than `minSimilarity` but has no nodes to be connected with (for example, by a certain filtering status). I've added an additional filtering process for this [here](https://github.com/hms-dbmi/cistrome-higlass-wrapper/blob/9a4094e0fc5fe2aeba1043183672b41ffaaca482/src/utils/select-rows.js#L51). 

Moreover, since `minSimilarity` filter should look into other `transformedRows`, I needed to rearrange a local variable, `filterInfos` in `select-rows.js`, to make the `minSimilarity` filter to be taken lastly ([here](https://github.com/hms-dbmi/cistrome-higlass-wrapper/blob/ca01b6dc605e98e2918fe8a454e07973611aa933/src/utils/select-rows.js#L18)). I think this part should be reviewed again once we show more than two distance-contained dendrograms, which I will add a corresponding issue for this.

I've also added tests for `select-rows.js` ([here](https://github.com/hms-dbmi/cistrome-higlass-wrapper/blob/ca01b6dc605e98e2918fe8a454e07973611aa933/src/utils/select-rows.spec.js#L70)) -- These tests were totally necessary to be included in this PR as filters especially for `tree` data are making the function further complex.

I've also resolved all other comments you've proposed.
